### PR TITLE
[BEAM-662] Fix for allowing floating point periods in windows

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @robertwb, @aaltay PTAL

### CODE REVIEW

**Comment by sb2nov** [comment]
Two main changes here:
1. We were calculating the end time from the start of the range instead of what the windows will the timestamp fall into
2. Using Duration/Timestamp directly into the range function causes a loss of precision

**Comment by aaltay** [comment]
Retest this please

File: sdks/python/apache_beam/transforms/window.py [hunk 1]
```diff
@@ -380,7 +380,7 @@ class SlidingWindows(NonMergingWindowFn):
     period: Period of the windows as seconds.
     offset: Offset of this window as seconds since Unix epoch. Windows start at
       t=N * period + offset where t=0 is the epoch. The offset must be a value
-      in range [0, period). If it is not it will be normalized to this range.
+      in range [0, size). If it is not it will be normalized to this range.
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window.py]: Period is correct here. We need to fix the code.

File: sdks/python/apache_beam/transforms/window.py [hunk 2]
```diff
@@ -392,9 +392,11 @@ def __init__(self, size, period, offset=0):
 
   def assign(self, context):
     timestamp = context.timestamp
-    start = timestamp - (timestamp - self.offset) % self.period
-    return [IntervalWindow(Timestamp.of(s), Timestamp.of(s) + self.size)
-            for s in range(start, start - self.size, -self.period)]
+    start = timestamp - ((timestamp - self.offset) % self.period)
+    return [
+        IntervalWindow(Timestamp(micros=s), Timestamp(micros=s) + self.size)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window.py]: Should we remove the truncating Timestamp.__int__ which allowed this bug?

Comment (sb2nov) [inline sdks/python/apache_beam/transforms/window.py]: We could, that does change the return type from seconds -> microseconds. Do you know if there is any other place that depends on it, I wasn't sure about that part.

File: sdks/python/apache_beam/transforms/window.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/window.py]: Thanks for catching. I fixed it.

File: sdks/python/apache_beam/transforms/window.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window.py]: I wouldn't change the return type, just remove it altogether. It just means one couldn't write int(timestamp) and get the (implicitly truncated) timestamps anymore, including using them in range. 

I don't think anyone is using it (and if they are, likely there's a bug like above).

Comment (sb2nov) [inline sdks/python/apache_beam/transforms/window.py]: Done.

File: sdks/python/apache_beam/transforms/window_test.py [hunk 3]
```diff
@@ -108,6 +108,20 @@ def test_sliding_windows_assignment(self):
...
+    self.assertEqual([IntervalWindow(1.5, 5.0), IntervalWindow(-1.0, 2.5)],
+                     windowfn.assign(context('v', 1.7)))
+    self.assertEqual([IntervalWindow(1.5, 5.0)],
+                     windowfn.assign(context('v', 3)))
+
+  def test_sliding_windows_assignment_fraction_large_offset(self):
+    windowfn = SlidingWindows(size=3.5, period=2.5, offset=4.0)
+    self.assertEqual([IntervalWindow(1.5, 5.0), IntervalWindow(-1.0, 2.5)],
+                     windowfn.assign(context('v', 1.7)))
+    self.assertEqual([IntervalWindow(1.5, 5.0)],
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/window_test.py]: Maybe choose a value like 4.5 that demonstrations a window starting at the requested offset of 4.0?

File: sdks/python/apache_beam/utils/timestamp.py [hunk 4]
```diff
@@ -163,14 +163,6 @@ def __repr__(self):
       return 'Duration(%s%d.%06d)' % (sign, int_part, frac_part)
     return 'Duration(%s%d)' % (sign, int_part)
 
-  def __float__(self):
```
Comment (robertwb) [inline sdks/python/apache_beam/utils/timestamp.py]: I think float is still useful, please restore.

File: sdks/python/apache_beam/transforms/window_test.py (hunk 3 above)
Comment (sb2nov) [inline sdks/python/apache_beam/transforms/window_test.py]: Done

File: sdks/python/apache_beam/utils/timestamp.py (hunk 4 above)
Comment (sb2nov) [inline sdks/python/apache_beam/utils/timestamp.py]: Done

**Pull request closed without merging**